pygfx/renderers/wgpu/meshshader.py

```diff
@@ -329,21 +346,34 @@ def code_fragment(self):
 
             // Lighting
             $$ if lighting
-                let world_pos = varyings.world_pos;
+                // Get view direction
                 let view = select(
-                    normalize(u_stdinfo.cam_transform_inv[3].xyz - world_pos),
+                    normalize(u_stdinfo.cam_transform_inv[3].xyz - varyings.world_pos),
                     ( u_stdinfo.cam_transform_inv * vec4<f32>(0.0, 0.0, 1.0, 0.0) ).xyz,
                     is_orthographic()
                 );
-                // Get normal
-                var normal = vec3<f32>(varyings.normal);
+                // Get surface normal
                 $$ if flat_shading
-                let u = dpdx(varyings.world_pos);
-                let v = dpdy(varyings.world_pos);
-                normal = normalize(cross(u, v));
-                normal = select(normal, -normal, (select(0, 1, is_front) + u_stdinfo.flipped_winding) == 1);
+                    let u = dpdx(varyings.world_pos);
+                    let v = dpdy(varyings.world_pos);
+                    var normal = normalize(cross(u, v));
+                    // Compute the normal from the 'world_pos' partial derivatives with respect to window coordinates.
+                    // Therefore no need to flip the normal by face orientation, but we need consider the sign of the camera dimension.
+                    normal = select(-normal, normal, varyings.winding_cam < 0.0);
+                $$ else
+                    $$ if use_normal_map is defined
+                        var normal_map = textureSample( t_normal_map, s_normal_map, varyings.texcoord );
+                        normal_map = normal_map * 2.0 - 1.0;
+                        let normal_map_scale = vec3<f32>( normal_map.xy * u_material.normal_scale, normal_map.z );
+                        normal = perturbNormal2Arb(view, normal, normal_map_scale, varyings.texcoord, is_front);
+                    $$ else
+                        var normal = vec3<f32>(varyings.normal);
+                    $$ endif
+                    // Flip the vertex normal if we're looking at the back face
+                    normal = select(-normal, normal, is_front);
```

Comment:
I refactored this to also do the flipping for backfaces when a normal map is supplied. Otherwise the back/inside of such meshes would be dark. @panxinmiao I think this makes sense, do you agree?

Reply:
Oh wait, this seems to be covered in `perturbNormal2Arb`. My bad. Will revert.

Reply:
Yes, the flipped `normal` is a parameter of `perturbNormal2Arb`.

Here is a test model to confirm the correctness of the processing: https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/NormalTangentTest

This is the test code: https://gist.github.com/panxinmiao/f4037a2a3bf037dfb476170a735bc76a

BTW, this also confirms a previous problem. When loading the pbr model from trimesh, the default value of `normal_scale` should be (1, - 1). I also corrected this problem in this PR.